memiavl/wal.go

```diff
@@ -0,0 +1,124 @@
+package memiavl
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/tidwall/wal"
+)
+
+type Change struct {
+	Delete     bool
+	Key, Value []byte
+}
+
+var (
+	DefaultPathToWAL string
+)
+
+const (
+	// amount of bytes it takes to store key's and value's lengths
+	KeyValueLen = 4
+)
+
+// ChangeBz is a byte slice containing information about one change to memiavl.
+// Structure is as follows:
+// deleteByte: 1 byte (0 or 1)
+// keyLen: 4 bytes (length of key)
+// key: key bytes
+
+// if delete is 0, also contains:
+
+// valueLen: 4 bytes (length of value)
+// value: value bytes
+type ChangeBz = []byte
+
+func init() {
+	userHomeDir, err := os.UserHomeDir()
+	if err != nil {
+		panic(err)
+	}
+
+	DefaultPathToWAL = filepath.Join(userHomeDir, ".cronosd/data/application.wal")
+}
+
+func NewWAL(pathToWAL string, opts *wal.Options) (*wal.Log, error) {
+	if pathToWAL == "" {
+		return nil, fmt.Errorf("failed trying to create a new WAL: path to WAL is empty")
+	}
+
+	return wal.Open(pathToWAL, opts)
+}
+
+// prepareChange prepares a change for write-ahead log.
+// If value provided is nil, it implies the Change should indicate deletion.
+func prepareChange(key []byte, value []byte) Change {
+	return Change{
+		Delete: value == nil,
+		Key:    key,
+		Value:  value,
+	}
+}
+
+// writeChange writes a change to the write-ahead log.
+// Returns bytes written
+func writeChange(wal *wal.Log, change Change, index uint64) (uint64, error) {
```

Comment:
I suggest storing the change set of a whole block as a single entry in wal, since that's the atomic unit to replay.

Reply:
makes sense! thx

Reply:
For simplicity we can reuse the [change set formats used in versiondb](https://github.com/crypto-org-chain/cronos/blob/bd7251b866bbe3944b997ea002ab846a0dccf6cb/versiondb/client/changeset.go#L31), in memory type would be `[]iavl.KVPair`

Reply:
https://github.com/crypto-org-chain/cronos/pull/952

You can build on top of this API, I've implemented the marshal/unmarshal for the `ChangeSet` type.

Reply:
Another thing is someone need to receive the raw `Set`/`Remove` calls and accumulate the change set, I think we can let the `KVStore` wrapper to do that, so memiavl only expose the `ApplyChangeSet` API, rather than the raw `Set/Remove` APIs.

Reply:
oh, I already implemented it but yeah no problem I'll merge your branch, thx for letting me know!

Reply:
I just did a little bit more changes, so now memiavl tree only expose a single method `ApplyChangeSet` to do modifications, I think you can merge the branch now, I just need some review from colleagues to merge.

Reply:
Just realize there's `iavl.ChangeSet` type already, so I changed the PR to reuse that type.

Reply:
#952 is merged.

Reply:
applied those changes to this PR